did-pkh/src/lib.rs

```diff
@@ -0,0 +1,715 @@
+use async_trait::async_trait;
+use chrono::prelude::*;
+
+use ssi::caip10::BlockchainAccountId;
+use ssi::did::{
+    Context, Contexts, DIDMethod, Document, Source, VerificationMethod, VerificationMethodMap,
+    DEFAULT_CONTEXT, DIDURL,
+};
+use ssi::did_resolve::{
+    DIDResolver, DocumentMetadata, ResolutionInputMetadata, ResolutionMetadata, ERROR_INVALID_DID,
+    TYPE_DID_LD_JSON,
+};
+use ssi::jwk::{Base64urlUInt, OctetParams, Params, JWK};
+
+// https://github.com/ChainAgnostic/CAIPs/blob/master/CAIPs/caip-4.md
+const CHAIN_ID_BITCOIN_MAINNET: &str = "bip122:000000000019d6689c085ae165831e93";
```

Comment:
I'm wondering if did-pkh should care about which network the address is on. Are there cases for the blockchain network being a higher level concern than cryptographic operations mangling a public key, simplifying this model? @awoie made good points in this thread, but would still like to hear about use cases or reasons to include the network. I'm not necessarily opposed. And if we do include network, should we consider just adopting the did-caip10 format all-together? How do we handle non-caip10 formats?

https://github.com/spruceid/ssi/issues/123#issuecomment-802111650

Reply:
The chain id is used here to construct the [`blockchainAccountId`](https://www.w3.org/TR/did-spec-registries/#blockchainaccountid) property to put in the verification method object. A verifier uses the value of the `blockchainAccountId` property to verify that the public key that signed the proof corresponds to the property value. Essentially we are using [`blockchainAccountId`](https://w3c-ccg.github.io/security-vocab/#blockchainAccountId) to encode the DID's key type and public key hash, without coupling the verification methods to the DID method. The verification methods can use `blockchainAccountId` without needing to know bout `did:pkh` in particular.

If we omit the chain id, then we are not using `blockchainAccountId` and must instead specify a new property to express the public key hash and type in the verification method. My assumption is that it is good for interoperability and separation of concerns for verification methods to try to stay decoupled from particular DID methods, although for generative DID methods they may end up looking similar.

I opened an issue on `EcdsaSecp256k1RecoverySignature2020` to discuss use of `blockchainAccountId` and possible generalization beyond it (e.g. for non-blockchain use cases): https://github.com/decentralized-identity/EcdsaSecp256k1RecoverySignature2020/issues/10. `EcdsaSecp256k1RecoverySignature2020` is the only verification method I am aware of that uses `blockchainAccountId`, besides our own `Ed25519BLAKE2BDigestSize20Base58CheckEncodedSignature2021` and `P256BLAKE2BDigestSize20Base58CheckEncodedSignature2021`.

I think making a `did:caip10` directly could be good, to make use of the existing `blockchainAccountId` property, and define less new things. A small downside is it would result in somewhat longer DIDs, due to needing the chain id for BIP122, as seen here. The larger downside is that it would be specific to blockchain projects, by definition. So far we only have blockchain use cases, but we could imagine going beyond that. I think CAIP is meant to support any blockchain, but maybe not cryptographic ids outside blockchains. `blockchainAccountId` can express a public key hash, but semantically it means blockchain account address rather than just some public key material.

Possible non-blockchain use cases for `did:pkh` (Each of these would probably need a new proof & VM type):
- SSH key fingerprints. <https://tools.ietf.org/html/rfc4255>.
- PGP key fingerprints. <https://tools.ietf.org/html/rfc4880#section-12.2>. This might make more sense as its own DID method using a key server for updates, but it could be useful for purely generative use as well.
- RSA keys - for when the public key is be too large to use with `did:key`.

Reply:
Makes sense to me, I'm in favor of these changes. It is not clear that CAIP-10 would be limited to PKH processed bytes, whereas this DID method's scope makes that limitation explicitly clear. Therefore, CAIP-10 is not necessarily a proper subset of PKH plus metadata, and we should not seek to even confine it under something like `did:pkh:caip10`. If someone wants to use CAIP-10 as the method-specific identifier, they should create a new method. Thanks for the great explanation.